test/migration/verify/verify_migration_test.go

```diff
@@ -0,0 +1,161 @@
+package verify
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/api/v1alpha1"
+	"github.com/codeready-toolchain/toolchain-common/pkg/states"
+	"github.com/codeready-toolchain/toolchain-e2e/test/migration"
+	. "github.com/codeready-toolchain/toolchain-e2e/testsupport"
+	"github.com/codeready-toolchain/toolchain-e2e/testsupport/cleanup"
+	"github.com/codeready-toolchain/toolchain-e2e/testsupport/md5"
+	"github.com/codeready-toolchain/toolchain-e2e/testsupport/wait"
+	"github.com/stretchr/testify/require"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestAfterMigration(t *testing.T) {
+	// given
+	awaitilities := WaitForDeployments(t)
+
+	// get Signups for the users provisioned in the setup part
+	provisionedSignup := getSignupFromMUR(t, awaitilities.Host(), migration.ProvisionedUser)
+	secondMemberProvisionedSignup := getSignupFromMUR(t, awaitilities.Host(), migration.SecondMemberProvisionedUser)
+	appstudioProvisionedSignup := getSignupFromMUR(t, awaitilities.Host(), migration.AppStudioProvisionedUser)
+	deactivatedSignup := listAndGetSignupWithState(t, awaitilities.Host(), toolchainv1alpha1.UserSignupStateLabelValueDeactivated)
+	bannedSignup := listAndGetSignupWithState(t, awaitilities.Host(), toolchainv1alpha1.UserSignupStateLabelValueBanned)
+
+	var wg sync.WaitGroup
+
+	// prepare all functions to verify the state of the Signups and Spaces
+	toRun := []func(){
+		// Spaces
+		func() { verifyAppStudioProvisionedSpace(t, awaitilities) },
+		func() { verifySecondMemberProvisionedSpace(t, awaitilities) },
+		// UserSignups
+		func() { verifyProvisionedSignup(t, awaitilities, provisionedSignup) },
+		func() { verifySecondMemberProvisionedSignup(t, awaitilities, secondMemberProvisionedSignup) },
+		func() { verifyAppStudioProvisionedSignup(t, awaitilities, appstudioProvisionedSignup) },
+		func() { verifyDeactivatedSignup(t, awaitilities, deactivatedSignup) },
+		func() { verifyBannedSignup(t, awaitilities, bannedSignup) },
+	}
+
+	// when & then - run all functions in parallel
+	for _, funcToRun := range toRun {
+		wg.Add(1)
+		go func(run func()) {
+			defer wg.Done()
+			run()
+		}(funcToRun)
+	}
+
+	wg.Wait()
+
+	cleanup.ExecuteAllCleanTasks()
+
+	t.Run("run migration setup", func(t *testing.T) {
+		// We need to run the migration setup part to be sure that when the PR is merged
```

Comment:
I appreciate the explanation but I don't understand it: how does running this subtest will help with upcoming PRs? On OpenShift CI, the next PRs will be tested on their own/isolated clusters, so what's the point of setting up migration resources here? 
I believe I'm totally missing the point, though :/

Reply:
Consider this:
1. there is running version of host/member-operator that populates the `UserAccount.Spec.NSTemplateSet` field
2. the `migration setup logic` checks and expects that this field is populated
3. there is a migration PR that drops usage of that `UserAccount.Spec.NSTemplateSet` field so it won't be populated any more
4. there is also paired e2e PR that modifies the `verify migration test` so it doesn't expect the field to be populated
5. the tests are green - so far so good.

but what would happen when the PRs are merged?
The next PR would fail in the `migration setup logic` because it would still contain the logic that would check and expect `UserAccount.Spec.NSTemplateSet` to be populated.

Reply:
ok, I see. In that case, the next PR will have to update the migration setup logic so that the field is not populated anymore. Or even more radical, it could be a PR with an API update where the field is removed: since the previous PR was already merged, we can assume it was deployed and that the migration took place, so we should be able to move on with the removal of the deprecated field.

Reply:
> In that case, the next PR will have to update the migration setup logic so that the field is not populated anymore.

but who should do that? imagine that in the meantime there can be multiple other PRs and for all of them the e2e tests would start failing. In addition, it would fail also for PRs in other repositories (host, member, reg-serv).
They would all need to wait until the migration setup logic is fixed. I believe that making the logic compatible with both versions is a much easier solution

> it could be a PR with an API update where the field is removed

removal of the deprecated field is a different story. I was talking about the logic that populates the field - so the content matters, not the presence.

Anyway, that was just an example. That use-case could fit to any situation where the setup part relies on a functionality/feature that is present in the current version of operators but won't be present in the next one.
So, to make our lives easier, let's assume that everything that is merged in master works as expected. This means that we can "just" create either `UserSignup` or `Space` and wait for the provisioned status (or any other status/state that signs that the process is done). We don't have to verify the actual content of the resources, nor labels, etc... because it was already verified when the PR was merged. 
If we agree on such a generic setup logic, then we can easily make sure that it's fully compatible with both versions of Dev Sandbox.

Reply:
I missed the point of this until I saw this discussion too. I guess that the other options (making the next PR update the migration logic) would lead to other problems that sound worse than keeping it this way. So my vote would be to keep this approach unless there's a better way. If we decide to keep it this way then maybe we should add Matous' explanation to the comment since it wasn't obvious to two people.

Reply:
The current approach seems to be reasonable to me too but I agree with Rajiv that it's better to add more details to the comment with the reasoning. Some generic example (similar to what you did here in this discussion) would be the best way of explaining the need.

Reply:
I updated the comment here 3046308
Feel free to improve/fix/change it if needed  :wink: